compiler/src/dotty/tools/dotc/typer/ConstFold.scala

```diff
@@ -89,6 +90,14 @@ object ConstFold:
     case (nme.UNARY_- , FloatTag  ) => Constant(-x.floatValue)
     case (nme.UNARY_- , DoubleTag ) => Constant(-x.doubleValue)
 
+    case (nme.toByte  , ByteTag | CharTag | ShortTag | IntTag | LongTag | FloatTag | DoubleTag ) => Constant(x.byteValue)
+    case (nme.toChar  , ByteTag | CharTag | ShortTag | IntTag | LongTag | FloatTag | DoubleTag ) => Constant(x.charValue)
+    case (nme.toShort , ByteTag | CharTag | ShortTag | IntTag | LongTag | FloatTag | DoubleTag ) => Constant(x.shortValue)
+    case (nme.toInt   , ByteTag | CharTag | ShortTag | IntTag | LongTag | FloatTag | DoubleTag ) => Constant(x.intValue)
+    case (nme.toLong  , ByteTag | CharTag | ShortTag | IntTag | LongTag | FloatTag | DoubleTag ) => Constant(x.longValue)
+    case (nme.toFloat , ByteTag | CharTag | ShortTag | IntTag | LongTag | FloatTag | DoubleTag ) => Constant(x.floatValue)
+    case (nme.toDouble, ByteTag | CharTag | ShortTag | IntTag | LongTag | FloatTag | DoubleTag ) => Constant(x.doubleValue)
```

Comment:
`case (nme.toByte, _) if x.isNumeric => ???` perhaps

Reply:
nice! thanks for the suggestion